### ISSUE TITLE
Config Singleton

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#1158]
It would be really nice if libvalhalla had a config singleton. Something where we initialize it once and then can access it globally after that. Why would we want that...

Right now we have the annoying problem of having to pass a config to all of the top level elements. It would be nice at program start to just configure the singleton and let these objects get access to the config on the fly.

One problem this would solve directly would be that we have config values that get validated against for requests. Currently a loki worker will load these when its instantiated and then validate each request against them. This makes it so that some of the things we'd like to pull out to shared request object cannot be because the code that genrates this pbf object doesnt have access to the config only loki_worker.

Basically having a globally configured valhalla is just convenient :smile: 

more details:

1. singleton getter takes an optional config file so that on first call it gets configured (i think the python bindings used to be like this, we'd just move this idea into the library)
2. the config object knows whether or not its configured. if its not configured we throw in the getter or maybe even assert to force kill the application? this forces who ever is using it to make sure their use includes configuring the library first.
3. wrap the config boilerplate that every excecutable will have into the arg parsing boilerplate that we are adding in #4169

**Comment by kevinkreiser** [comment]
taking a bit on here to also remove all ptree dependencies at least in the main library (sans executables, tests, and data creation)

**Comment by kevinkreiser** [comment]
we found a way to use ptree for now in header only mode. so im rescoping this back to config singleton

**Comment by janusz-anue** [comment]
When the singleton is implemented, should it replace every config-variable that was passed before with a getter to the singleton? I.e:
```cpp
void run_service(const boost::property_tree::ptree& config) {
  // gracefully shutdown when asked via SIGTERM
  prime_server::quiesce(config.get<unsigned int>("httpd.service.drain_seconds", 28),
                        config.get<unsigned int>("httpd.service.shutting_seconds", 1));
  // ...
}
```

to (for example)

```cpp
void run_service() {
  auto config = valhalla::config::GetConfig();

  // gracefully shutdown when asked via SIGTERM
  prime_server::quiesce(config.get<unsigned int>("httpd.service.drain_seconds", 28),
                        config.get<unsigned int>("httpd.service.shutting_seconds", 1));
  // ...
}
```

**Comment by nilsnolde** [comment]
That's how I'd imagine it, yep. I know it's toooons of code changes, but it'll be superb when done, thanks for doing this! You might've guessed/seen it already, but we don't care too much about C++ API (in)stability:)

**Comment by kevinkreiser** [comment]
we can do it in phases. you can just let all the ptree arguments as they are and just call the singleton before hand at the top level. that would reduce the amount of change. then later we can remove all the ptree passing and have iseres of the config call the singleton directly

**Comment by kevinkreiser** [comment]
also just for the sake of brevity and inline-ness we could make it a free function in the valhalla namespace

```c++
inline const boost::property_tree::ptree& config(const std::string& config_location_or_inline);
```

this has the benifit of making it so succinct in most places that people will just do

```c++
config().get<unsigned int>("some.property");
```

and another benefit that the name config will likely collide with any local variables named as such, cluing you into the places you need to update. it could also be a terrible idea but it just might be crazy enough to work :smile:
